pkg/stripe/analytics_telemetry.go

```diff
@@ -0,0 +1,169 @@
+type TelemetryClient interface {
+	SendAPIRequestEvent(ctx context.Context, requestID string, livemode bool) (*http.Response, error)
+	SendEvent(ctx context.Context, eventName string, eventValue string) (*http.Response, error)
+}
+
+// AnalyticsTelemetryClient sends event information to r.stripe.com
+type AnalyticsTelemetryClient struct {
+	BaseURL    *url.URL
+	WG         *sync.WaitGroup
+	HTTPClient *http.Client
+}
+
+//
+// Public functions
+//
+
+// InitContext initializes an instance of CLIAnalyticsEventContext
+func InitContext() *CLIAnalyticsEventContext {
+	return &CLIAnalyticsEventContext{
+		InvocationID: uuid.NewString(),
+		UserAgent:    useragent.GetEncodedUserAgent(),
+		CLIVersion:   version.Version,
+		OS:           runtime.GOOS,
+	}
+}
+
+// SetCobraCommandContext sets the telemetry values for the command being executed.
+func (e *CLIAnalyticsEventContext) SetCobraCommandContext(cmd *cobra.Command) {
+	e.CommandPath = cmd.CommandPath()
+	e.GeneratedResource = false
+
+	for _, value := range cmd.Annotations {
+		// Generated commands have an annotation called "operation", we can
+		// search for that to let us know it's generated
+		if value == "operation" {
+			e.GeneratedResource = true
+		}
+	}
+}
+
+// SetMerchant sets the merchant on the CLIAnalyticsEventContext object
+func (e *CLIAnalyticsEventContext) SetMerchant(merchant string) {
+	e.Merchant = merchant
+}
+
+// SendAPIRequestEvent is a special function for API requests
+func (a *AnalyticsTelemetryClient) SendAPIRequestEvent(ctx context.Context, requestID string, livemode bool) (*http.Response, error) {
+	a.WG.Add(1)
+	defer a.WG.Done()
+	if (ctx.Value(TelemetryContextKey{}) != nil) {
+		data, _ := query.Values(ctx.Value(TelemetryContextKey{}))
+
+		data.Set("client_id", "stripe-cli")
+		data.Set("request_id", requestID)
+		data.Set("livemode", strconv.FormatBool(livemode))
+		data.Set("event_id", uuid.NewString())
+		data.Set("event_name", "API Request")
+		data.Set("event_value", "")
+		data.Set("created", fmt.Sprint((time.Now().Unix())))
+
+		return a.sendData(ctx, data)
+	}
+	return nil, nil
+}
+
+// SendEvent sends a telemetry event to r.stripe.com
+func (a *AnalyticsTelemetryClient) SendEvent(ctx context.Context, eventName string, eventValue string) (*http.Response, error) {
+	a.WG.Add(1)
+	defer a.WG.Done()
+	if (ctx.Value(TelemetryContextKey{}) != nil) {
+		data, _ := query.Values(ctx.Value(TelemetryContextKey{}))
+
+		data.Set("client_id", "stripe-cli")
+		data.Set("event_id", uuid.NewString())
+		data.Set("event_name", eventName)
+		data.Set("event_value", eventValue)
+		data.Set("created", fmt.Sprint((time.Now().Unix())))
+
+		return a.sendData(ctx, data)
+	}
+	return nil, nil
+}
+
+func (a *AnalyticsTelemetryClient) sendData(ctx context.Context, data url.Values) (*http.Response, error) {
+	if telemetryOptedOut(os.Getenv("STRIPE_CLI_TELEMETRY_OPTOUT")) {
```

Comment:
I think this would be less surprising if this check was moved into `main.go`, so that the context only contains a telemetry client if we're supposed to be sending telemetry. Since the `sendTelemetryEvent` below skips sending events if the context doesn't contain the `TelemetryClientKey` key, I think it would "just work"?

That would also address your concern below about not being happy about setting env vars inside the tests -- instead you could test each case by either adding the client to the context or not.

Reply:
I like it, updated!